public/lesswrong-editor/src/ckeditor5-footnote/src/utils.js

```diff
@@ -0,0 +1,225 @@
+// @ts-check
+import ModelElement from '@ckeditor/ckeditor5-engine/src/model/element';
+import ModelText from '@ckeditor/ckeditor5-engine/src/model/text';
+import ModelTextProxy from '@ckeditor/ckeditor5-engine/src/model/textproxy';
+import ViewElement from '@ckeditor/ckeditor5-engine/src/view/element';
+import ViewText from '@ckeditor/ckeditor5-engine/src/view/text';
+import ViewTextProxy from '@ckeditor/ckeditor5-engine/src/view/textproxy';
+import { Editor } from '@ckeditor/ckeditor5-core';
+
+// There's ample DRY violation in this file; type checking
+// polymorphism without full typescript is just incredibly finicky.
+// I suspect there's a more elegant solution for this, 
+// but I tried a lot of things and none of them worked.
+
+/**
+ * Returns an array of all descendant elements of
+ * the root for which the provided predicate returns true.
+ * @param {Editor} editor
+ * @param {ModelElement} rootElement  
+ * @param {(item: ModelElement) => boolean} predicate
+ * @returns {ModelElement[]} */
+ export const modelQueryElementsAll = (
```

Comment:
It looks to me like you are supposed to use a [TreeWalker](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_model_treewalker-TreeWalker.html) for this? Something like `new TreeWalker({startPosition: rootElement}).skip(x => !predicate(x))`